app/experimenter/static/js/design-page.js

```diff
@@ -0,0 +1,494 @@
+          <FormLabel>
+            <strong>Design</strong>
+          </FormLabel>
+          <br />
+          <a href="/">help</a>
+        </Col>
+        <Col md={9}>
+          <FormControl
+            name="design"
+            type="textarea"
+            value={props.design}
+            onChange={props.handleInputChange}
+          />
+        </Col>
+      </Row>
+    </div>
+  );
+}
+
+function TypeForm(props) {
+  const type = props.type;
+
+  if (type == "pref") {
+    return <PrefForm {...props}/>;
+  } else if (type == "addon") {
+    return <AddonForm {...props}/>;
+  } else {
+    return <GenericForm {...props}/>;
+  }
+}
+
+export default class DesignForm extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.slug = props.slug
+
+    this.addBranch = this.addBranch.bind(this);
+    this.removeBranch = this.removeBranch.bind(this);
+
+    this.updateRatio = this.updateRatio.bind(this);
+    this.updateName = this.updateName.bind(this);
+    this.updateDescription = this.updateDescription.bind(this);
+    this.updateValue = this.updateValue.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.handleInputChange = this.handleInputChange.bind(this);
+
+    this.state = {
+      type: props.expType,
+      pref_key: "",
+      pref_type: "",
+      pref_branch: "",
+      addon_experiment_id: "",
+      addon_release_url: "",
+      design: "",
+      variants: [{
+        ratio: null,
+        name: "",
+        description: "",
+        value: "",
+        is_control: true,
+      }, {
+        ratio: null,
+        name: "",
+        description: "",
+        value: "",
+        is_control: false,
+      }],
+      errors: {}
+    };
+  }
+
+  componentDidMount() {
+    const that = this;
+
+    let url;
+    if (this.state.type == "pref") {
+      url = `/api/v1/experiments/${this.slug}/design-pref`
+    } else if (this.state.type == "addon") {
+      url = `/api/v1/experiments/${this.slug}/design-addon`
+    } else {
+      url = `/api/v1/experiments/${this.slug}/design-generic`
+    }
+
+    fetch(url)
+      .then(function(response) {
+        return response.json();
+      })
+      .then(function(json) {
+        if (json.variants == "") {
+          return that.setState({
+            type: json.type,
+            pref_key: json.pref_key,
+            pref_type: json.pref_type,
+            pref_branch: json.pref_branch,
+            addon_experiment_id: json.addon_experiment_id,
+            addon_release_url:json.addon_release_url,
+            design: json.design
+          })
+        } else {
+          return that.setState({
+            type: json.type,
+            pref_key: json.pref_key,
+            pref_type: json.pref_type,
+            pref_branch: json.pref_branch,
+            addon_experiment_id: json.addon_experiment_id,
+            addon_release_url:json.addon_release_url,
+            design: json.design,
+            variants: json.variants
```

Comment:
If the only difference between these two cases is this one field then can you just create the object once and optionally add this field before you pass it to `setState`?

Reply:
spread operator is your friend

Reply:
I can just get rid of all of this repetition by setting the state to the json just like... 

`  return that.setState(json)`

this one line is a huge improvement on like 30 repeating ones.

Reply:
(and fixing the that issue also).